### ISSUE TITLE
Quirk: Sequence version of print_helper is called for c-string literals

### ISSUE THREAD

**Issue by james-perretta** [eecs280staff/unit_test_framework#30]
I stumbled across this while setting up a demo AG project. May or may not be worth changing.

Test case:
```
#include <string>
#include "unit_test_framework.h"

TEST(string_eq) {
    ASSERT_EQUAL("spam", string("egg"));
}

TEST_MAIN()
```

Output: 
```
** Test case "string_eq": FAIL
In test , line 17: 
{ s, p, a, m,  } != egg
```

**Comment by amirkamil** [comment]
A fix would necessarily assume that all `char` arrays are null-terminated. The following does the trick:

```
diff --git a/unit_test_framework.h b/unit_test_framework.h
index b976675..38bf9b6 100644
--- a/unit_test_framework.h
+++ b/unit_test_framework.h
@@ -198,9 +198,11 @@ auto print_helper(std::ostream& os, const Sequence& seq, int, ...)
     return os << " }";
 }
 
-// This version of print_helper will be called when T is an array.
+// This version of print_helper will be called when T is a non-char array.
 template <class Elem, std::size_t N>
-std::ostream& print_helper(std::ostream& os, const Elem (&arr)[N], int, int) {
+auto print_helper(std::ostream& os, const Elem (&arr)[N], int, int)
+    -> typename std::enable_if<!std::is_same<Elem, char>::value,
+                               std::ostream&>::type {
     if (N == 0) {
         return os << "{}";
     }
```

What say you?

**Comment by james-perretta** [comment]
I think that's a reasonable assumption. The `<<` operator makes the same assumption, no?

**Comment by amirkamil** [comment]
Assuming that a `char` array is null-terminated would make it UB to test non-null-terminated `char` arrays for equality with `ASSERT_SEQUENCE_EQUAL`:
```
    char arr1[] = {'s', 'p', 'a', 'm'};
    char arr2[] = {'e', 'g', 'g'};
    ASSERT_SEQUENCE_EQUAL(arr1, arr2);
```
This results in UB upon failure, e.g.:
```
** Test case "char_array_uneq": FAIL
In test , line 48: 
spam@??? != eggspam@???
```

Alternative suggestion: when presented with a `char` array, check for a null terminator and print as a string if present. Otherwise print as a sequence:
```
+// This version of print_helper will be called when T is a char array.
+// If the array contains a null terminator, it is printed as a string.
+// Otherwise, it is printed as a sequence.
+template <std::size_t N>
+std::ostream& print_helper(std::ostream& os, const char (&arr)[N], int, int) {
+    for (std::size_t i = 0; i < N; ++i) {
+        if (!arr[i]) {
+            return os << arr;
+        }
+    }
+    return print_sequence_helper(os, arr);
+}
```

Thoughts on this?

**Comment by james-perretta** [comment]
That's a good point. I think your suggested alternative has a similar problem: If the char arrays passed to `ASSERT_SEQUENCE_EQUAL` happen to have null terminating characters at position `N`, then they'll print at strings, which is puzzling at best.

Would it be worth adding a dedicated `ASSERT_CSTRING_EQUAL` with a requirement that the arguments be C-strings?

**Comment by amirkamil** [comment]
All of the assertions use the same printing functions, and I think that is the right choice to properly handle nested structures. For `char` arrays, we have to make a choice about whether they are printed as arrays or strings. The former is the status quo. My most recent proposal is to print them as strings if they contain a null terminator anywhere (not just at the end), but as arrays if not. I believe this will handle most cases as expected while avoiding UB.

The question of adding an `ASSERT_CSTRING_EQUAL` is an orthogonal one -- right now, we do not have an assertion that can directly compare two `char *`'s as C-style strings. It might also be worth adding another assertion that takes a custom comparator.

**Comment by james-perretta** [comment]
Ah I see, I misread your example code. I agree that your most recent proposal will produce the desired behavior in the motivating example in my original post.

With regard to `print_sequence_helper`: would the contents of the sequence version of `print_helper` be moved into `print_sequence_helper`?

It may also be worth asking at this point: is adding another `print_helper` overload worth it when the user can work around the issue by using `string("spam")`, or in c++14 the even shorter `"spam"s`?

**Comment by amirkamil** [comment]
> With regard to print_sequence_helper: would the contents of the sequence version of `print_helper` be moved into `print_sequence_helper`?

Yes, and all of the sequence, array, and `char` array overloads would invoke `print_sequence_helper`.

> It may also be worth asking at this point: is adding another `print_helper` overload worth it when the user can work around the issue by using `string("spam")`, or in c++14 the even shorter `"spam"s`?

I don't have a strong opinion on this. I defer to your opinion as the OP for this issue.

**Comment by amirkamil** [comment]
Fixed by PR #35.
